cratedb_mcp/knowledge.py

```diff
@@ -100,3 +109,78 @@ class Queries:
         "link": "https://raw.githubusercontent.com/crate/cratedb-guide/9ab661997d7704ecbb63af9c3ee33535957e24e6/docs/performance/optimization.rst"
     }
 ]
+
+
+def sql_expression_permitted(expression: str) -> bool:
+    """
+    Validate the SQL expression, only permit read queries by default.
+
+    When the `CRATEDB_MCP_PERMIT_ALL_STATEMENTS` environment variable is set,
+    allow all types of statements.
+
+    FIXME: Revisit implementation, it might be too naive or weak.
+           Issue:    https://github.com/crate/cratedb-mcp/issues/10
+           Question: Does SQLAlchemy provide a solid read-only mode, or any other library?
+    """
+    outcome = _sql_expression_permitted(expression)
+    if outcome is True:
+        logger.info(f"Permitted SQL expression: {expression}")
+    else:
+        logger.warning(f"Denied SQL expression: {expression}")
+    return outcome
+
+
+def _sql_expression_permitted(expression: str) -> bool:
+
+    if expression:
+        expression = expression.strip()
+
+    if not expression:
+        return False
+
+    if PERMIT_ALL_STATEMENTS:
+        return True
+
+    # Parse the SQL statement.
+    parsed = cratedb_sqlparse.sqlparse(expression)
+
+    # Reject multiple statements to prevent potential SQL injections.
+    if len(parsed) > 1:
+        return False
+
+    # Check if the expression is valid and if it's a SELECT statement,
+    # also trying to consider `SELECT ... INTO ...` statements.
+    operation = parsed[0].type.upper()
+    is_select = operation == 'SELECT'
+    is_rejected = _sql_is_select_into(expression) or _sql_is_evasive(expression)
+    if is_select and not is_rejected:
+        return True
+    return False
+
+
+def _sql_is_select_into(expression: str) -> bool:
+    """
+    Helper function using traditional `sqlparse` for catching `SELECT ... INTO ...` statements.
+
+    Examples:
+
+        SELECT * INTO foobar FROM bazqux
+        SELECT * FROM bazqux INTO foobar
+    """
+    parsed = sqlparse.parse(expression)
+    tokens = [str(item).upper() for item in parsed[0]]
+    return "INTO" in tokens
+
+
+def _sql_is_evasive(expression: str) -> bool:
+    """
+    Helper function using traditional `sqlparse` for catching evasive SQL statements.
+
+    Reject multiple statements to prevent potential SQL injections.
+
+    Examples:
+
+        SELECT * FROM users; \uff1b DROP TABLE users
+    """
+    parsed = sqlparse.parse(expression)
+    return len(parsed) > 1
```

Comment:
I had to add those two helper functions, still using traditional `sqlparse`, to catch those statements. I haven't been able to do it with `cratedb-sqlparse`, so a few test cases were failing.

Reply:
See GH-14.